File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.go
/*
Time - O(n)
Space - O(1)
*/
func maxProfit(prices []int) int {
    buy:=0
    sell:=1
    maxProfit:=0
    for sell<len(prices) {
        profit:=prices[sell]-prices[buy]
        if profit>=0 {
            if profit>maxProfit {
                maxProfit=profit
            }
            sell++
        }else {
            buy=sell
            sell++
        }
    }
    return maxProfit
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy=0
        sell=1
        n=len(prices)
        maxProfit=0
        while sell<n:
            profit=prices[sell]-prices[buy]
            if profit<0:
                buy=sell
            else:
                maxProfit=max(maxProfit, profit)
            sell+=1
        return maxProfit
        